File: CSharp/One/Compiler.cs
using One.Ast;
using One.Transforms;
using One;
using Parsers;
using StdLib;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace One
{
    public interface ICompilerHooks {
        void afterStage(string stageName);
    }
    
    public class Compiler {
        public PackageManager pacMan;
        public Workspace workspace;
        public SourceFile nativeFile;
        public ExportedScope nativeExports;
        public Package projectPkg;
        public ICompilerHooks hooks;
        
        public Compiler()
        {
            this.pacMan = null;
            this.workspace = null;
            this.nativeFile = null;
            this.nativeExports = null;
            this.projectPkg = null;
            this.hooks = null;
        }
        
        public async Task init(string packagesDir)
        {
            this.pacMan = new PackageManager(new PackagesFolderSource(packagesDir));
            await this.pacMan.loadAllCached();
        }
        
        public ITransformer[] getTransformers(bool forDeclarationFile)
        {
            var transforms = new List<ITransformer>();
            if (forDeclarationFile) {
                transforms.push(new FillParent());
                transforms.push(new FillAttributesFromTrivia());
                transforms.push(new ResolveImports(this.workspace));
                transforms.push(new ResolveGenericTypeIdentifiers());
                transforms.push(new ResolveUnresolvedTypes());
                transforms.push(new FillMutabilityInfo());
            }
            else {
                transforms.push(new FillParent());
                transforms.push(new FillAttributesFromTrivia());
                transforms.push(new ResolveImports(this.workspace));
                transforms.push(new ResolveGenericTypeIdentifiers());
                transforms.push(new ConvertToMethodCall());
                transforms.push(new ResolveUnresolvedTypes());
                transforms.push(new ResolveIdentifiers());
                transforms.push(new InstanceOfImplicitCast());
                transforms.push(new DetectMethodCalls());
                transforms.push(new InferTypes());
                transforms.push(new CollectInheritanceInfo());
                transforms.push(new FillMutabilityInfo());
                transforms.push(new LambdaCaptureCollector());
            }
            return transforms.ToArray();
        }
        
        public void setupNativeResolver(string content)
        {
            this.nativeFile = TypeScriptParser2.parseFile(content);
            this.nativeExports = Package.collectExportsFromFile(this.nativeFile, true);
            foreach (var trans in this.getTransformers(true))
                trans.visitFiles(new SourceFile[] { this.nativeFile });
        }
        
        public void newWorkspace(string pkgName = "@")
        {
            this.workspace = new Workspace();
            foreach (var intfPkg in this.pacMan.interfacesPkgs) {
                var libName = $"{intfPkg.interfaceYaml.vendor}.{intfPkg.interfaceYaml.name}-v{intfPkg.interfaceYaml.version}";
                this.addInterfacePackage(libName, intfPkg.definition);
            }
            
            this.projectPkg = new Package(pkgName, false);
            this.workspace.addPackage(this.projectPkg);
        }
        
        public void addInterfacePackage(string libName, string definitionFileContent)
        {
            var libPkg = new Package(libName, true);
            var file = TypeScriptParser2.parseFile(definitionFileContent, new SourcePath(libPkg, Package.INDEX));
            this.setupFile(file);
            libPkg.addFile(file, true);
            this.workspace.addPackage(libPkg);
        }
        
        public void setupFile(SourceFile file)
        {
            file.addAvailableSymbols(this.nativeExports.getAllExports());
            file.literalTypes = new LiteralTypes((((Class)file.availableSymbols.get("TsBoolean"))).type, (((Class)file.availableSymbols.get("TsNumber"))).type, (((Class)file.availableSymbols.get("TsString"))).type, (((Class)file.availableSymbols.get("RegExp"))).type, (((Class)file.availableSymbols.get("TsArray"))).type, (((Class)file.availableSymbols.get("TsMap"))).type, (((Class)file.availableSymbols.get("Error"))).type, (((Class)file.availableSymbols.get("Promise"))).type);
            file.arrayTypes = new ClassType[] { (((Class)file.availableSymbols.get("TsArray"))).type, (((Class)file.availableSymbols.get("IterableIterator"))).type, (((Class)file.availableSymbols.get("RegExpExecArray"))).type, (((Class)file.availableSymbols.get("TsString"))).type, (((Class)file.availableSymbols.get("Set"))).type };
        }
        
        public void addProjectFile(string fn, string content)
        {
            var file = TypeScriptParser2.parseFile(content, new SourcePath(this.projectPkg, fn));
            this.setupFile(file);
            this.projectPkg.addFile(file);
        }
        
        public void processFiles(SourceFile[] files)
        {
            foreach (var trans in this.getTransformers(false)) {
                trans.visitFiles(files);
                if (this.hooks != null)
                    this.hooks.afterStage(trans.name);
            }
        }
        
        public void processWorkspace()
        {
            foreach (var pkg in Object.values(this.workspace.packages).filter(x => x.definitionOnly))
                foreach (var trans in this.getTransformers(true))
                    trans.visitFiles(Object.values(pkg.files));
            
            this.processFiles(Object.values(this.projectPkg.files));
        }
    }
}

File: PHP/One/Compiler.php
<?php

namespace One\Compiler;

use One\Ast\Types\Workspace;
use One\Ast\Types\Package;
use One\Ast\Types\SourcePath;
use One\Ast\Types\SourceFile;
use One\Ast\Types\ExportedScope;
use One\Ast\Types\LiteralTypes;
use One\Ast\Types\Class_;
use One\Ast\Types\ExportScopeRef;
use Parsers\TypeScriptParser2\TypeScriptParser2;
use StdLib\PackageManager\PackageManager;
use StdLib\PackagesFolderSource\PackagesFolderSource;
use One\Transforms\FillParent\FillParent;
use One\Transforms\FillAttributesFromTrivia\FillAttributesFromTrivia;
use One\Transforms\ResolveGenericTypeIdentifiers\ResolveGenericTypeIdentifiers;
use One\Transforms\ResolveUnresolvedTypes\ResolveUnresolvedTypes;
use One\Transforms\ResolveImports\ResolveImports;
use One\Transforms\ConvertToMethodCall\ConvertToMethodCall;
use One\Transforms\ResolveIdentifiers\ResolveIdentifiers;
use One\Transforms\InstanceOfImplicitCast\InstanceOfImplicitCast;
use One\Transforms\DetectMethodCalls\DetectMethodCalls;
use One\Transforms\InferTypes\InferTypes;
use One\Transforms\CollectInheritanceInfo\CollectInheritanceInfo;
use One\Transforms\FillMutabilityInfo\FillMutabilityInfo;
use One\AstTransformer\AstTransformer;
use One\ITransformer\ITransformer;
use One\Transforms\LambdaCaptureCollector\LambdaCaptureCollector;

interface ICompilerHooks {
    function afterStage($stageName);
}

class Compiler {
    public $pacMan;
    public $workspace;
    public $nativeFile;
    public $nativeExports;
    public $projectPkg;
    public $hooks;
    
    function __construct()
    {
        $this->pacMan = null;
        $this->workspace = null;
        $this->nativeFile = null;
        $this->nativeExports = null;
        $this->projectPkg = null;
        $this->hooks = null;
    }
    
    function init($packagesDir) {
        $this->pacMan = new PackageManager(new PackagesFolderSource($packagesDir));
        $this->pacMan->loadAllCached();
    }
    
    function getTransformers($forDeclarationFile) {
        $transforms = array();
        if ($forDeclarationFile) {
            $transforms[] = new FillParent();
            $transforms[] = new FillAttributesFromTrivia();
            $transforms[] = new ResolveImports($this->workspace);
            $transforms[] = new ResolveGenericTypeIdentifiers();
            $transforms[] = new ResolveUnresolvedTypes();
            $transforms[] = new FillMutabilityInfo();
        }
        else {
            $transforms[] = new FillParent();
            $transforms[] = new FillAttributesFromTrivia();
            $transforms[] = new ResolveImports($this->workspace);
            $transforms[] = new ResolveGenericTypeIdentifiers();
            $transforms[] = new ConvertToMethodCall();
            $transforms[] = new ResolveUnresolvedTypes();
            $transforms[] = new ResolveIdentifiers();
            $transforms[] = new InstanceOfImplicitCast();
            $transforms[] = new DetectMethodCalls();
            $transforms[] = new InferTypes();
            $transforms[] = new CollectInheritanceInfo();
            $transforms[] = new FillMutabilityInfo();
            $transforms[] = new LambdaCaptureCollector();
        }
        return $transforms;
    }
    
    function setupNativeResolver($content) {
        $this->nativeFile = TypeScriptParser2::parseFile($content);
        $this->nativeExports = Package::collectExportsFromFile($this->nativeFile, true);
        foreach ($this->getTransformers(true) as $trans)
            $trans->visitFiles(array($this->nativeFile));
    }
    
    function newWorkspace($pkgName = "@") {
        $this->workspace = new Workspace();
        foreach ($this->pacMan->interfacesPkgs as $intfPkg) {
            $libName = $intfPkg->interfaceYaml->vendor . "." . $intfPkg->interfaceYaml->name . "-v" . $intfPkg->interfaceYaml->version;
            $this->addInterfacePackage($libName, $intfPkg->definition);
        }
        
        $this->projectPkg = new Package($pkgName, false);
        $this->workspace->addPackage($this->projectPkg);
    }
    
    function addInterfacePackage($libName, $definitionFileContent) {
        $libPkg = new Package($libName, true);
        $file = TypeScriptParser2::parseFile($definitionFileContent, new SourcePath($libPkg, Package::$INDEX));
        $this->setupFile($file);
        $libPkg->addFile($file, true);
        $this->workspace->addPackage($libPkg);
    }
    
    function setupFile($file) {
        $file->addAvailableSymbols($this->nativeExports->getAllExports());
        $file->literalTypes = new LiteralTypes(($file->availableSymbols->get("TsBoolean"))->type, ($file->availableSymbols->get("TsNumber"))->type, ($file->availableSymbols->get("TsString"))->type, ($file->availableSymbols->get("RegExp"))->type, ($file->availableSymbols->get("TsArray"))->type, ($file->availableSymbols->get("TsMap"))->type, ($file->availableSymbols->get("Error"))->type, ($file->availableSymbols->get("Promise"))->type);
        $file->arrayTypes = array(($file->availableSymbols->get("TsArray"))->type, ($file->availableSymbols->get("IterableIterator"))->type, ($file->availableSymbols->get("RegExpExecArray"))->type, ($file->availableSymbols->get("TsString"))->type, ($file->availableSymbols->get("Set"))->type);
    }
    
    function addProjectFile($fn, $content) {
        $file = TypeScriptParser2::parseFile($content, new SourcePath($this->projectPkg, $fn));
        $this->setupFile($file);
        $this->projectPkg->addFile($file);
    }
    
    function processFiles($files) {
        foreach ($this->getTransformers(false) as $trans) {
            $trans->visitFiles($files);
            if ($this->hooks !== null)
                $this->hooks->afterStage($trans->name);
        }
    }
    
    function processWorkspace() {
        foreach (array_values(array_filter(array_values($this->workspace->packages), function ($x) { return $x->definitionOnly; })) as $pkg)
            foreach ($this->getTransformers(true) as $trans)
                $trans->visitFiles(array_values($pkg->files));
        
        $this->processFiles(array_values($this->projectPkg->files));
    }
}
